drizzlepac/hlautils/product.py

```diff
@@ -272,10 +303,35 @@ def wcs_drizzle_product(self, meta_wcs):
                                   output=self.drizzle_filename,
                                   **drizzle_pars)
 
+        # Update product with SVM-specific keywords based on the footprint
+        print("\n\nMASK_KWS: \n{}\n\n".format(self.mask_kws))
+        with fits.open(self.drizzle_filename, mode='update') as hdu:
+            for kw in self.mask_kws:
+                hdu[("SCI", 1)].header[kw] = tuple(self.mask_kws[kw])
+
         # Rename Astrodrizzle log file as a trailer file
         log.debug("Filter combined image {} composed of: {}".format(self.drizzle_filename, edp_filenames))
         shutil.move(self.trl_logname, self.trl_filename)
 
+    def generate_footprint_mask(self):
+        """ Create a footprint mask for a set of exposure images
+
+            Create a mask which is True/1/on for the illuminated portion of the image, and
+            False/0/off for the remainder of the image.
+        """
+        footprint = cell_utils.SkyFootprint(self.meta_wcs)
+        exposure_names = [element.full_filename for element in self.edp_list]
+        footprint.build(exposure_names, scale=True)
+        self.mask = footprint.total_mask
+
+        # Compute footprint-based SVM-specific keywords for product image header
+        good_pixels = self.mask > 0
+        self.mask_kws['NPIXFRAC'][0] = good_pixels.sum()
+        self.mask_kws['MEANEXPT'][0] = np.mean(footprint.scaled_mask[good_pixels])
+        self.mask_kws['MEDEXPT'][0] = np.median(footprint.scaled_mask[good_pixels])
+        self.mask_kws['MEANNEXP'][0] = np.mean(self.mask[good_pixels])
+        self.mask_kws['MEDNEXP'][0] = np.median(self.mask[good_pixels])
+
```

Comment:
It looks like _generate_footprint_mask()_  and _wcs_drizzle_product()_ are basically the same for the _TotalProduct_ and _FIlterProduct_ classes.  It is clear you do not want the _ExposureProduct_ class to have these methods.  You could put these method in the _HAPProduct_ base class and then override the methods in the _ExposureProduct_ class with "pass" which is a cheapo solution. Alternatively, you could do what is discussed at this URL which I found compelling:
https://medium.com/@maouu/sorry-but-youre-wrong-aea1b88ffc03

Reply:
The alternative definitions using __getattribute__ and __dict__ for ExposureProduct was implemented successfully(?) in 871bdef.  Thanks for the pointer!